File: algorithms/counting-sort/c/countingsort.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int find_max(int *a, int n)
{
  int max = a[0];
  for (int i = 1; i < n; i++) {
    if (a[i] > max) {
      max = a[i];
    }
  }
  return max;
}

void countingsort(int *a, int n)
{
  int k = find_max(a, n) + 1;
  int *c = (int *)malloc(sizeof(int) * k);
  int *t = (int *)malloc(sizeof(int) * n);

  if (c == NULL || t == NULL) {
    return;
  }

  memset(c, 0, k);

  for (int i = 0; i < n; i++) {
    c[a[i]]++;
  }

  for (int i = 1; i < k; i++) {
    c[i] += c[i - 1];
  }

  for (int i = n - 1; i >= 0; i--) {
    t[--c[a[i]]] = a[i];
  }

  for (int i = 0; i < n; i++) {
    a[i] = t[i];
  }

  free(c);
  free(t);
}

File: algorithms/counting-sort/python/countingsort.py
#!/usr/bin/env python3

def find_max(a, n):
  max = a[0]
  for i in range(1, n):
    if a[i] > max:
      max = a[i]
  return max

def countingsort(a):
  n = len(a)
  k = find_max(a, n) + 1

  c = [0] * k
  t = [0] * n

  for i in range(n):
    c[a[i]] += 1

  for i in range(1, k):
    c[i] += c[i - 1]

  for i in range(n - 1, -1, -1):
    c[a[i]] -= 1
    t[c[a[i]]] = a[i]

  for i in range(n):
    a[i] = t[i]

  return a

